src/libsyntax/ast.rs

```diff
@@ -1254,6 +1254,23 @@ pub enum ExprKind {
     Err,
 }
 
+impl ExprKind {
+    /// Whether this expression can appear in a contst argument without being surrounded by braces.
+    ///
+    /// Only used in error recovery.
+    pub(crate) fn is_valid_const_on_its_own(&self) -> bool {
+        match self {
+            ExprKind::Tup(_) |
```

Comment:
I think understanding this is subtle enough that some argumentation around each case would be good with a comment, using "example" style.